alertmanager/peers.go

```diff
@@ -0,0 +1,87 @@
+package alertmanager
+
+import (
+	"net"
+	"time"
+
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/common/log"
+)
+
+var (
+	srvRequestDuration = prometheus.NewHistogramVec(prometheus.HistogramOpts{
+		Namespace: "cortex",
+		Name:      "srv_lookup_request_duration_seconds",
+		Help:      "Time spent looking up SRV records.",
+		Buckets:   prometheus.DefBuckets,
+	}, []string{"service", "proto", "hostname", "status_code"})
+)
+
+func init() {
+	prometheus.MustRegister(srvRequestDuration)
+}
+
+// TODO: change memcache_client to use this.
+
+// SRVDiscovery discovers SRV services.
+type SRVDiscovery struct {
+	Service      string
+	Proto        string
+	Hostname     string
+	PollInterval time.Duration
+	Addresses    chan []*net.SRV
+
+	stop chan struct{}
+	done chan struct{}
+}
+
+// NewSRVDiscovery makes a new SRVDiscovery.
+func NewSRVDiscovery(service, hostname string, pollInterval time.Duration) *SRVDiscovery {
+	disco := &SRVDiscovery{
+		Service:      service,
+		Proto:        "tcp",
+		Hostname:     hostname,
+		PollInterval: pollInterval,
+		Addresses:    make(chan []*net.SRV),
+		stop:         make(chan struct{}),
+		done:         make(chan struct{}),
+	}
+	go disco.loop()
+	return disco
+}
+
+// Stop the SRVDiscovery
+func (s *SRVDiscovery) Stop() {
+	close(s.stop)
+	<-s.done
+}
+
+func (s *SRVDiscovery) updatePeers() {
+	var addrs []*net.SRV
```

Comment:
I would keep all of this simpler:

- have two plain counters (total / failure) instead of full histogram with fake HTTP-style status code dimension
- hardcode the service and proto dimensions and remove them from the metrics (probably also remove hostname dimension from the metric if it's always the same within the process)